### DISCUSSION TITLE
Energy usage

### DISCUSSION THREAD

**Discussion by RonnyRusten** [robbinjanssen/home-assistant-ojmicroline-thermostat#72] [General]
Is there any way to get the energy usage from the MVD5?
I saw the discussion on https://community.home-assistant.io/t/mwd5-wifi-thermostat-oj-electronics-microtemp/445601, but it didn't conclude.

**Comment by robbinjanssen** [comment]
I think it should be possbile, however I currently have no time to research/implement it. Feel free to give it a go! :-)

**Reply by RonnyRusten** [reply]
If I had the knowledge of how to do it, I certainly would have tried. Unfortunately I don't.

**Reply by RonnyRusten** [reply]
OK, I'm very new to this, but I have found that the data is fetched in a file called ojmicroline.py.
If I'm reading this correctly it's not part of your integration.
Can I edit this file? 
Sorry if this is a stupid question, as I said it's the first time I try this.
I have set up a dev container and "installed" the integration as described in the manual setup section.
-edit-
I found the repo for [python-ojmicroline-thermostat](https://github.com/robbinjanssen/python-ojmicroline-thermostat), so now I'm a little less confused...

**Reply by robbinjanssen** [reply]
Here's a quick "how does it work"

The library you found is used to actually talk with the API of the thermostat.
This integration uses that library to get the data. This is an abstraction layer which is recommended when creating a home assistant integration.

So for example [here](https://github.com/robbinjanssen/python-ojmicroline-thermostat/blob/main/ojmicroline_thermostat/ojmicroline.py#L178-L208) a call is made to "get" the information from the ojmicroline API and parse them into [Thermostat](https://github.com/robbinjanssen/python-ojmicroline-thermostat/blob/main/ojmicroline_thermostat/models/thermostat.py#L27) models. This makes it easier to use the data in this integration.

In the "coordinator" of this integration (the one responsible for "handling data updates"). It creates a [connection](https://github.com/robbinjanssen/home-assistant-ojmicroline-thermostat/blob/main/custom_components/ojmicroline_thermostat/coordinator.py#L42-L49) with the python-ojmicroline-thermostat library and then calls the [`get_thermostats`](https://github.com/robbinjanssen/home-assistant-ojmicroline-thermostat/blob/main/custom_components/ojmicroline_thermostat/coordinator.py#L68-L69) method.

Then the integrations creates [sensors](https://github.com/robbinjanssen/home-assistant-ojmicroline-thermostat/blob/main/custom_components/ojmicroline_thermostat/sensor.py#L38-L45) based on each model (Device) and basically calls the [method](https://github.com/robbinjanssen/home-assistant-ojmicroline-thermostat/blob/main/custom_components/ojmicroline_thermostat/sensor.py#L164-L173) on the Thermostat model (you can read `getattr(self.coordinator.data[self.idx], self.entity_description.key)` as `thermostat.temperature_room`)

---

Now when we want to do something with the Energy usage, the first thing that should happen is create methods in the `python-ojmicroline-thermostat` package to actually get the data from the thermostat/API and parse it.

**Reply by RonnyRusten** [reply]
Thank you for the explanation, and it makes sense, indeed. I kind of thought this was how it works, but thank you for confirming.
Now I need to find out how I run and debug the `python-ojmicroline-thermostat` package, but that is probably easy.
I have never done anything in python before, so I need to google a lot (everything) :-)

So, if I'm able to do anything useful in `python-ojmicroline-thermostat`, I should make a PR on that, then you confirm and update (hopefully). THEN I can do "stuff" in the integration?
I read somewhere (https://developers.home-assistant.io/docs/creating_integration_manifest#requirements) that these libraries are installed by HA automatically, but I guess it's possible to force it in "manually" if I'm impatient or want to test things before I bother you?

If you have any links to usefull things I should read before asking a lot of silly questions, please let me know :-)

**Reply by robbinjanssen** [reply]
You are on the right path! :-)

If you "fork" the python-ojmicroline-thermostat repository you can make changes and later create a pull request to get them into the "main" branch. There is a [`test_output.py`](https://github.com/robbinjanssen/python-ojmicroline-thermostat/blob/main/test_output.py#L40-L46) file which should get you going.

When done you can indeed modify the integration (you can do this in your home assistant instance as well, doesn't really mather).. When you want to use _your_ version of python-ojmicroline-thermostat, you can indeed modify the [`maniftest`](https://github.com/robbinjanssen/home-assistant-ojmicroline-thermostat/blob/main/custom_components/ojmicroline_thermostat/manifest.json#L11-L13) file and point it to your forked repository :)

**Reply by RonnyRusten** [reply]
OK, I'll see if I can make this work.
Thanks!

BTW:
The CustomerId. It is set to 99, but shouldn't it be set to my ID? I found a number in the app, but I haven't tried to just leave it at 99...

**Reply by robbinjanssen** [reply]
Do you already use the integration in home assistant? Then use the same details as filled in there :-)

It all depends on which app you use on your phone to control the thermostats. For example I use the SWATT app and it forces me to use customer ID 99.. (i've managed to retrieve this information by sniffing the network requests, which is a pain in the ass to get done)..

More details: https://github.com/robbinjanssen/home-assistant-ojmicroline-thermostat/discussions/56

**Reply by RonnyRusten** [reply]
OK. I use the Microtemp app (Android) and in my thermostat information I have a customer ID. I have used this in the integration and it works, but I will try to see what happens if I use a different ID :-)
![2023-08-30_22-35-14](https://github.com/robbinjanssen/home-assistant-ojmicroline-thermostat/assets/1955294/770a1340-0435-4136-8d3b-66e3ded1266d)

**Reply by robbinjanssen** [reply]
Then you should definitely use that customer ID :-) just modify the test_output.py file and see if it works for you.

**Comment by TorStava** [comment]
Thank you again for your work in this component.

I have submitted pull request #289 to implement energy usage for the OWD5/MWD5 thermostat. It requires that the accompanying python library release version >= 3.3.0.

**Comment by Tockra** [comment]
Hey, can somebody explain what the energy_usage is showing?
I checked the sensor and it seems to reset sometimes. I don't think I can use this sensor for energy dashboard, because it assumes to have a total sum sensor:
![grafik](https://github.com/user-attachments/assets/e2ec4649-7e1a-4a20-9d80-e5b43b010136)
